examples/bots/discord_bot.py

```diff
@@ -0,0 +1,193 @@
+    """
+
+    def __init__(
+        self,
+        chat_agent: ChatAgent,
+        channel_ids: Optional[List[int]] = None,
+        discord_token: Optional[str] = None,
+        auto_retriever: Optional[AutoRetriever] = None,
+        content_input_paths: Union[str, List[str]] = "",
+        top_k: int = 1,
+        return_detailed_info: bool = True,
+    ) -> None:
+        self.chat_agent = chat_agent
+        self.token = discord_token or os.getenv('DISCORD_TOKEN')
+        self.channel_ids = channel_ids
+        self.auto_retriever = auto_retriever
+        self.content_input_paths = content_input_paths
+        self.top_k = top_k
+        self.return_detailed_info = return_detailed_info
+
+        if not self.token:
+            raise ValueError(
+                "`DISCORD_TOKEN` not found in environment variables. Get it"
+                " here: `https://discord.com/developers/applications`."
+            )
+
+        try:
+            import discord
+        except ImportError:
+            raise ImportError(
+                "Please install `discord` first. You can install it by running"
+                " `python3 -m pip install -U discord.py`."
+            )
+        intents = discord.Intents.default()
+        intents.message_content = True
+        self.client = discord.Client(intents=intents)
+
+        # Register event handlers
+        self.client.event(self.on_ready)
+        self.client.event(self.on_message)
+
+    def run(self) -> None:
+        r"""Start the Discord bot using its token.
+
+        This method starts the Discord bot by running the client with the
+        provided token.
+        """
+        self.client.run(self.token)  # type: ignore[arg-type]
+
+    async def on_ready(self) -> None:
+        r"""This method is called when the bot has successfully connected to
+        the Discord server.
+
+        It prints a message indicating that the bot has logged in and displays
+        the username of the bot.
+        """
+        print(f'We have logged in as {self.client.user}')
+
+    async def on_message(self, message: 'Message') -> None:
+        r"""Event handler for when a message is received.
+
+        Args:
+            message (discord.Message): The message object received.
+        """
+
+        # If the message author is the bot itself,
+        # do not respond to this message
+        if message.author == self.client.user:
+            return
+
+        # If allowed channel IDs are provided,
+        # only respond to messages in those channels
+        if self.channel_ids and message.channel.id not in self.channel_ids:
+            return
+
+        # Only respond to messages that mention the bot
+        if not self.client.user or not self.client.user.mentioned_in(message):
+            return
+
+        self.chat_agent.reset()
+
+        user_raw_msg = message.content
+
+        if self.auto_retriever:
+            user_raw_msg = self.auto_retriever.run_vector_retriever(
```

Comment:
here we need to write some prompt to re-organize the content with user query and content retrieved. like:

"Here is the query to you: {user_raw_msg}
Based on the retrieved content:{retrieved_content}, answer the query from {message.author.name}"